### PR TITLE
Wfst concurency

### PR DESCRIPTION
**Pull request by index-git**
Part of issue #159

- [x] Tests
- ~~Layman Test Client~~
- [ ] Changlelog
- [ ] Documentation

Optionally
- Add dependency to doc/dependencies.md

### CODE REVIEW

File: src/layman/celery.py [hunk 1]
```diff
@@ -29,6 +31,12 @@ def task_postrun(workspace, publication_type, publication_name, task_id, task_na
     hash = task_id
     if rds.hexists(key, hash):
         finnish_publication_task(task_id)
+        next_task = pop_step_to_run_after_chain(workspace, publication_type, publication_name)
+        if next_task:
+            module_name, method_name = next_task.split('::')
+            module = importlib.import_module(module_name)
+            method = getattr(module, method_name, None)
```
Comment (jirik) [inline src/layman/celery.py:38]: `, None` is redundant

File: src/layman/celery.py [hunk 2]
```diff
@@ -40,6 +48,30 @@ def _get_task_hash(task_name, workspace, publication_name):
     return f"{task_name}:{workspace}:{publication_name}"
 
 
+def push_step_to_run_after_chain(workspace, publication_type, publication_name, step_code, ):
+    rds = settings.LAYMAN_REDIS
+    key = LAST_TASK_ID_IN_CHAIN_TO_PUBLICATION
+    hash = _get_publication_hash(workspace, publication_type, publication_name)
+    val = rds.hget(key, hash)
+    queue = json.loads(val) if val is not None else list()
+    queue.append(step_code)
```
Comment (jirik) [inline src/layman/celery.py:57]: Should append only if it's not already there

File: src/layman/celery.py [hunk 3]
```diff
@@ -8,6 +9,7 @@
 REDIS_CURRENT_TASK_NAMES = f"{__name__}:CURRENT_TASK_NAMES"
 PUBLICATION_CHAIN_INFOS = f'{__name__}:PUBLICATION_TASK_INFOS'
 LAST_TASK_ID_IN_CHAIN_TO_PUBLICATION = f'{__name__}:TASK_ID_TO_PUBLICATION'
+RUN_AFTER_CHAIN = f'{__name__}:RUN_AFTER_CHAIN'
```
Comment (jirik) [inline src/layman/celery.py:12]: Should be cleared in abort_chain

File: src/layman/celery.py [hunk 4]
```diff
@@ -40,6 +48,30 @@ def _get_task_hash(task_name, workspace, publication_name):
...
+    rds = settings.LAYMAN_REDIS
+    key = LAST_TASK_ID_IN_CHAIN_TO_PUBLICATION
+    hash = _get_publication_hash(workspace, publication_type, publication_name)
+    val = rds.hget(key, hash)
+    result = None
+    if val:
+        queue = json.loads(val)
+        if len(queue) > 0:
+            result = queue.pop(0)
+        rds.hset(key, hash, json.dumps(queue))
```
Comment (jirik) [inline src/layman/celery.py:71]: `rds.hset(key, hash, json.dumps(queue))` should be inside second `if`

File: src/layman/celery.py (hunk 1 above)
Comment (jirik) [reply src/layman/celery.py:38]: Maybe move to another commit

File: src/layman/common/redis.py [hunk 5]
```diff
@@ -83,13 +83,12 @@ def solve_locks(workspace, publication_type, publication_name, error_code, metho
         if method in [consts.REQUEST_METHOD_PATCH, consts.REQUEST_METHOD_POST, ]:
```
Comment (jirik) [inline src/layman/common/redis.py:83]: Add WFST ti the list

File: src/layman/error_list.py [hunk 6]
```diff
@@ -49,4 +49,5 @@
     46: (400, 'Unknown style file. Can recognize only SLD and QML files.'),
     47: (400, 'Error in QML'),
     48: (400, 'Wrong combination of parameters'),
+    49: (400, 'Publication already in process'),
```
Comment (jirik) [inline src/layman/error_list.py:52]: Remove

File: src/layman/common/redis.py [hunk 7]
```diff
@@ -83,13 +83,12 @@ def solve_locks(workspace, publication_type, publication_name, error_code, metho
...
             raise LaymanError(error_code)
     if method not in [consts.REQUEST_METHOD_DELETE, ]:
-        if (current_lock, method) == (consts.REQUEST_METHOD_WFST, consts.REQUEST_METHOD_WFST):
+        if method == consts.REQUEST_METHOD_WFST:
+            raise LaymanError(49)
```
Comment (jirik) [inline src/layman/common/redis.py:87]: Raise 19 with special parameter `"can_run_later": True` of private_data

File: src/layman/geoserver_proxy.py [hunk 8]
```diff
@@ -197,8 +197,15 @@ def proxy(subpath):
 
     for workspace, layername in wfs_t_layers:
         if authz.can_i_edit(LAYER_TYPE, workspace, layername):
-            redis.create_lock(workspace, LAYER_TYPE, layername, 19, consts.REQUEST_METHOD_WFST)
-            patch_after_wfst(workspace, layername)
+            try:
+                redis.create_lock(workspace, LAYER_TYPE, layername, 19, consts.REQUEST_METHOD_WFST)
+                patch_after_wfst(workspace, layername)
+            except LaymanError as exc:
+                if exc.code == 49:
```
Comment (jirik) [inline src/layman/geoserver_proxy.py:204]: check 19 code and private_data.can_run_later

File: src/layman/common/redis.py [hunk 9]
```diff
@@ -83,13 +83,12 @@ def solve_locks(workspace, publication_type, publication_name, error_code, metho
...
+        if current_lock == consts.REQUEST_METHOD_WFST and method in [consts.REQUEST_METHOD_PATCH, ]:
```
Comment (jirik) [inline src/layman/common/redis.py:88]: Add also POST to the list

File: src/layman/requests_concurrency_test.py [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+    assert r.status_code == 200, r.text
+
+    r = requests.post(rest_url,
+                      data=data_xml,
+                      headers=headers)
+    assert r.status_code == 200, r.text
+
+    process_client.delete_workspace_layer(workspace, layer, )
+
+#
```
Comment (jirik) [inline src/layman/requests_concurrency_test.py]: Delete

File: src/layman/requests_concurrency_test.py [hunk 11]
```diff
@@ -1,9 +1,8 @@
-import time
+from test import process_client
+from test.data import wfs as data_wfs
```
Comment (jirik) [inline src/layman/requests_concurrency_test.py:2]: Fixup this commit

**Review by jirik — CHANGES_REQUESTED** [review]

**Comment by index-git** [comment]
Replaced with #359

**Pull request closed without merging**